classes/search/cmsfield.php

```diff
@@ -0,0 +1,231 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace mod_cms\search;
+
+defined('MOODLE_INTERNAL') || die();
+
+require_once($CFG->dirroot . '/mod/cms/lib.php');
+
+/**
+ * Define search area.
+ *
+ * @package    mod_cms
+ * @author     Tomo Tsuyuki <tomotsuyuki@catalyst-au.com>
+ * @copyright  2024 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class cmsfield extends \core_search\base_mod {
+
+    /**
+     * @var array Internal quick static cache.
+     */
+    protected $cmsdata = [];
+
+    /**
+     * @var array Internal quick static cache.
+     */
+    protected $defaultvalues = null;
+
+    /**
+     * Returns recordset containing required data for indexing cmsfield records.
+     *
+     * @param int $modifiedfrom timestamp
+     * @param \context|null $context Optional context to restrict scope of returned results
+     * @return \moodle_recordset|null Recordset (or null if no results)
+     */
+    public function get_document_recordset($modifiedfrom = 0, \context $context = null) {
+        global $DB;
+
+        list ($contextjoin, $contextparams) = $this->get_context_restriction_sql(
+                $context, 'cms', 'mc');
+        if ($contextjoin === null) {
+            return null;
+        }
+
+        // Search area is from customfield_data, but if the record is missing from activity, use default value.
+        $sql = "SELECT mc.id, mc.course AS courseid, mc.typeid, mcf.name AS fieldname, mcf.type,
+                       mcd.id dataid, mcd.value AS value, mcd.valueformat AS valueformat,
+                       mcd.timecreated AS timecreated, mcd.timemodified AS timemodified
+                  FROM {cms} mc
+                  JOIN {customfield_data} mcd ON mc.id = mcd.instanceid
+                  JOIN {customfield_field} mcf ON mcf.id = mcd.fieldid
+                  JOIN {customfield_category} mcc ON mcf.categoryid = mcc.id
+          $contextjoin
+                 WHERE mcd.timemodified >= ? AND mcc.component = 'mod_cms' AND mcc.area = 'cmsfield'
+                   AND mcf.type IN ('textarea', 'text')
+                 UNION
```

Comment:
with using union, wouldn't we get [mc.id](http://mc.id/) duplicated in the first column?

Reply:
Yes, it is possible.
I updated the SQL to concatenate contents, so both contents can be searched.